apps/speed.c

```diff
@@ -561,52 +639,514 @@ static OPT_PAIR ecdh_choices[] = {
 };
 #endif
 
+#ifndef SIGALRM
+# define COND(d) (count < (d))
+# define COUNT(d) (d)
+#else
+# define COND(c) (run && count<0x7fffffff)
+# define COUNT(d) (count)
+#endif                         /* SIGALRM */
+
+static unsigned char *buf = NULL, *buf2 = NULL;
+static int j;
+static char *engine_id = NULL;
+
+
+#ifndef OPENSSL_NO_MD2
+static unsigned char md2[MD2_DIGEST_LENGTH];
+static int EVP_Digest_MD2_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD2][j]); count++)
+        EVP_Digest(buf, (unsigned long)lengths[j], &(md2[0]), NULL,
+                EVP_md2(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MDC2
+static unsigned char mdc2[MDC2_DIGEST_LENGTH];
+static int EVP_Digest_MDC2_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MDC2][j]); count++)
+        EVP_Digest(buf, (unsigned long)lengths[j], &(mdc2[0]), NULL,
+                EVP_mdc2(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MD4
+static unsigned char md4[MD4_DIGEST_LENGTH];
+static int EVP_Digest_MD4_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD4][j]); count++)
+        EVP_Digest(&(buf[0]), (unsigned long)lengths[j], &(md4[0]),
+                NULL, EVP_md4(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MD5
+static unsigned char md5[MD5_DIGEST_LENGTH];
+static int MD5_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD5][j]); count++)
+        MD5(buf, lengths[j], md5);
+    return count;
+}
+
+static unsigned char hmac[MD5_DIGEST_LENGTH];
+static HMAC_CTX *hctx = NULL;
```

Comment:
This surely needs to be local? Otherwise you will reuse the same hctx across multiple jobs.